src/benchmark_runner.cc

```diff
@@ -140,23 +140,102 @@ void RunInThread(const BenchmarkInstance* b, IterationCount iters,
   manager->NotifyThreadComplete();
 }
 
+struct BenchTimeType {
+  enum { ITERS, TIME } tag;
+  union {
+    int iters;
+    double time;
+  };
+};
+
+BenchTimeType ParseBenchMinTime(const std::string& value) {
+  BenchTimeType ret;
+
+  if (value.empty()) {
+    ret.tag = BenchTimeType::TIME;
+    ret.time = 0.0;
+    return ret;
+  }
+
+  if (value.back() == 'x') {
+    std::string num_iters_str = value.substr(0, value.length() - 1);
+    int num_iters = std::atoi(num_iters_str.c_str());
+
+    // std::atoi doesn't provide useful error messages, so we do some
+    // sanity checks.
+    // Also check that the iters is not negative.
+    BM_CHECK(num_iters > 0 && !(num_iters == 0 && num_iters_str != "0"))
+        << "Malformed iters value passed to --benchmark_min_time: `" << value
+        << "`. Expected --benchmark_min_time=<integer>x.";
+
+    ret.tag = BenchTimeType::ITERS;
+    ret.iters = num_iters;
+    return ret;
+  }
+
+  std::string min_time_str;
+  if (value.back() != 's') {
+    BM_VLOG(2) << "Value passed to --benchmark_min_time should have a suffix. "
+                  "Eg., `30s` for 30-seconds.";
+    min_time_str = value;
+  } else {
+    min_time_str = value.substr(0, value.length() - 1);
+  }
+  double min_time = std::atof(min_time_str.c_str());
+
+  BM_CHECK(min_time > 0 &&
+           !(min_time == 0 && (min_time_str != "0" || min_time_str != "0.0")))
+      << "Malformed seconds value passed to --benchmark_min_time: `" << value
+      << "`. Expected --benchmark_min_time=<float>x.";
+
+  ret.tag = BenchTimeType::TIME;
+  ret.time = min_time;
+
+  return ret;
+}
+
+double GetMinTime(const benchmark::internal::BenchmarkInstance& b) {
+  if (!IsZero(b.min_time())) return b.min_time();
+  BenchTimeType iters_or_time = ParseBenchMinTime(FLAGS_benchmark_min_time);
+  // If the flag was used to specify number of iters, then return 0 for time.
+  if (iters_or_time.tag == BenchTimeType::ITERS) return 0;
+
+  return iters_or_time.time;
+}
+
+bool BenchMinTimeHasIters() {
```

Comment:
why does this go back to the flag if we already parsed it to determine the BenchTimeType?